### DISCUSSION TITLE
Best way to implement Dynamic DAG?

### DISCUSSION THREAD

**Discussion by mat7JT** [apache/airflow#32242] [Q&A]
Hello, I'm writing because I couldn't find any solution here or in the documentation.

**Briefly**, I need to write a DAG that can change its topology based on user input from the web UI.

More details: I have a workflow that takes a list of strings as input, which contains job names. It makes a single HTTP request to extract the ID associated with each job name, and finally, it executes a specific action for each of these IDs. I'm trying to implement this behavior by reading the job names from something that the web UI user can dynamically set. This input should also be permanent since the DAG will be scheduled.

Currently, I'm using an Airflow variable with a JSON object. Within my DAG, I read the variable (Variable.get), perform the HTTP request (SimpleHttpOperator), extract the IDs from the response (PythonOperator), and set them to a new variable (easier to use than the XCom response). In the DAG context, I extract the IDs from the new variable and finally create one operator for each ID (AirbyteTriggerSyncOperator). Basically, the number of these operators depends on the input.

When I update the variable, the DAG does not automatically re-draw until the first run is triggered. The first run breaks due to the DAG changes, but the next run finishes successfully.

What I have tried:
- Wrapping the dynamic part into a TaskGroup.
- Using a @task annotated method and invoking it with "expand(id=ids)".
- Wrapping the dynamic part into a subdag.

All of these attempts lead to a first run that either fails or ignores the input update. Do you have a solution (or even a resource) to suggest? 
Any hints would be greatly appreciated. Thank you.

**Comment by potiuk** [comment]
You want to use https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dynamic-task-mapping.html#dynamic-task-mapping 

It was specifically added to handle cases like yours

**Comment by mat7JT** [comment]
Thanks for your response. I have already tried using the dynamic task mapping approach (the "Using a @task annotated method ..." point). I apologize if I was not clear.

I have a quick update for the community: I have successfully managed to make it work by incorporating a slight delay of 30 seconds just before the dynamic section of my DAG. This delay is implemented using a PythonOperator that executes a sleep function. As a result, my DAG redraws successfully while the "wait operator" is being executed.
